crates/cdk/src/nuts/nut17.rs

```diff
@@ -0,0 +1,358 @@
+    fn from(proof_state: ProofState) -> NotificationPayload {
+        NotificationPayload::ProofState(proof_state)
+    }
+}
+
+impl From<MeltQuoteBolt11Response> for NotificationPayload {
+    fn from(melt_quote: MeltQuoteBolt11Response) -> NotificationPayload {
+        NotificationPayload::MeltQuoteBolt11Response(melt_quote)
+    }
+}
+
+impl From<MintQuoteBolt11Response> for NotificationPayload {
+    fn from(mint_quote: MintQuoteBolt11Response) -> NotificationPayload {
+        NotificationPayload::MintQuoteBolt11Response(mint_quote)
+    }
+}
+
+impl Indexable for NotificationPayload {
+    type Type = (String, Kind);
+
+    fn to_indexes(&self) -> Vec<Index<Self::Type>> {
+        match self {
+            NotificationPayload::ProofState(proof_state) => {
+                vec![Index::from((proof_state.y.to_hex(), Kind::ProofState))]
+            }
+            NotificationPayload::MeltQuoteBolt11Response(melt_quote) => {
+                vec![Index::from((
+                    melt_quote.quote.clone(),
+                    Kind::Bolt11MeltQuote,
+                ))]
+            }
+            NotificationPayload::MintQuoteBolt11Response(mint_quote) => {
+                vec![Index::from((
+                    mint_quote.quote.clone(),
+                    Kind::Bolt11MintQuote,
+                ))]
+            }
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy, Eq, Ord, PartialOrd, PartialEq, Hash, Serialize, Deserialize)]
+#[serde(rename_all = "snake_case")]
+
+/// Kind
+pub enum Kind {
+    /// Bolt 11 Melt Quote
+    Bolt11MeltQuote,
+    /// Bolt 11 Mint Quote
+    Bolt11MintQuote,
+    /// Proof State
+    ProofState,
+}
+
+impl AsRef<SubId> for Params {
+    fn as_ref(&self) -> &SubId {
+        &self.id
+    }
+}
+
+impl From<Params> for Vec<Index<(String, Kind)>> {
+    fn from(val: Params) -> Self {
+        let sub_id: SubscriptionGlobalId = Default::default();
+        val.filters
+            .iter()
+            .map(|filter| Index::from(((filter.clone(), val.kind), val.id.clone(), sub_id)))
+            .collect()
+    }
+}
+
+/// Manager
+#[derive(Default)]
+/// Publish–subscribe manager
+///
+/// Nut-17 implementation is system-wide and not only through the WebSocket, so
+/// it is possible for another part of the system to subscribe to events.
+pub struct PubSubManager(pub_sub::Manager<NotificationPayload, (String, Kind)>);
+
+impl Deref for PubSubManager {
+    type Target = pub_sub::Manager<NotificationPayload, (String, Kind)>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+#[cfg(feature = "mint")]
+impl From<&MintQuote> for MintQuoteBolt11Response {
+    fn from(mint_quote: &MintQuote) -> MintQuoteBolt11Response {
+        MintQuoteBolt11Response {
+            quote: mint_quote.id.clone(),
+            request: mint_quote.request.clone(),
+            state: mint_quote.state,
+            expiry: Some(mint_quote.expiry),
+        }
+    }
+}
+
```

Comment:
We have this in NUT04 lets move this there.


https://github.com/cashubtc/cdk/blob/d9fb5f814a745098c8a3c4ac8784ce2d28f52139/crates/cdk/src/nuts/nut04.rs#L95-L105